src/engine/LeakyReluConstraint.cpp

```diff
@@ -0,0 +1,583 @@
+/*********************                                                        */
+/*! \file LeakyReluConstraint.cpp
+ ** \verbatim
+ ** Top contributors (to current version):
+ **   Andrew Haoze Wu
+ ** This file is part of the Marabou project.
+ ** Copyright (c) 2017-2019 by the authors listed in the file AUTHORS
+ ** in the top-level source directory) and their institutional affiliations.
+ ** All rights reserved. See the file COPYING in the top-level source
+ ** directory for licensing information.\endverbatim
+ **
+ ** [[ Add lengthier description here ]]
+ **/
+
+#include "LeakyReluConstraint.h"
+
+#include "ConstraintBoundTightener.h"
+#include "ContextDependentPiecewiseLinearConstraint.h"
+#include "Debug.h"
+#include "DivideStrategy.h"
+#include "FloatUtils.h"
+#include "GlobalConfiguration.h"
+#include "ITableau.h"
+#include "InputQuery.h"
+#include "MStringf.h"
+#include "PiecewiseLinearCaseSplit.h"
+#include "MarabouError.h"
+#include "Statistics.h"
+#include "TableauRow.h"
+
+#ifdef _WIN32
+#define __attribute__(x)
+#endif
+
+LeakyReluConstraint::LeakyReluConstraint( unsigned b, unsigned f )
+    : ContextDependentPiecewiseLinearConstraint( TWO_PHASE_PIECEWISE_LINEAR_CONSTRAINT )
+    , _b( b )
+    , _f( f )
+    , _slope( GlobalConfiguration::LEAKY_RELU_SLOPE )
+    , _direction( PHASE_NOT_FIXED )
+    , _haveEliminatedVariables( false )
+{
+    if ( _slope <= 0 )
+        throw MarabouError( MarabouError::INVALID_LEAKY_RELU_SLOPE );
+}
+
+LeakyReluConstraint::LeakyReluConstraint( unsigned b, unsigned f, double slope )
+    : ContextDependentPiecewiseLinearConstraint( TWO_PHASE_PIECEWISE_LINEAR_CONSTRAINT )
+    , _b( b )
+    , _f( f )
+    , _slope( slope )
+    , _direction( PHASE_NOT_FIXED )
+    , _haveEliminatedVariables( false )
+{
+    if ( _slope <= 0 )
+        throw MarabouError( MarabouError::INVALID_LEAKY_RELU_SLOPE );
+}
+
+LeakyReluConstraint::LeakyReluConstraint( const String &serializedLeakyRelu )
+    : _haveEliminatedVariables( false )
+{
+    String constraintType = serializedLeakyRelu.substring( 0, 10 );
+    ASSERT( constraintType == String( "leaky_relu" ) );
+
+    // Remove the constraint type in serialized form
+    String serializedValues = serializedLeakyRelu.substring( 11, serializedLeakyRelu.length() - 11 );
+    List<String> values = serializedValues.tokenize( "," );
+
+    ASSERT( values.size() == 3 );
+
+    auto var = values.begin();
+    _f = atoi( var->ascii() );
+    ++var;
+    _b = atoi( var->ascii() );
+    ++var;
+    _slope = atof( var->ascii() );
+    if ( _slope <= 0 )
+        throw MarabouError( MarabouError::INVALID_LEAKY_RELU_SLOPE );
+
```

Comment:
can we add support for auxiliary variables? Without them, case splitting entails adding rows to the tableau, which we are trying to avoid.

Reply:
Done